sqld/src/replication/replica/injector/mod.rs

```diff
@@ -0,0 +1,239 @@
+        {
+            // create the replication table if it doesn't exist. We need to do that without hooks.
+            let connection = sqld_libsql_bindings::Connection::open(
+                path,
+                OpenFlags::SQLITE_OPEN_READ_WRITE
+                    | OpenFlags::SQLITE_OPEN_CREATE
+                    | OpenFlags::SQLITE_OPEN_URI
+                    | OpenFlags::SQLITE_OPEN_NO_MUTEX,
+                &TRANSPARENT_METHODS,
+                // safety: hook is dropped after connection
+                (),
+                DEFAULT_AUTO_CHECKPOINT,
+            )?;
+
+            connection.execute("CREATE TABLE IF NOT EXISTS libsql_temp_injection (x)", ())?;
+        }
+
+        let connection = sqld_libsql_bindings::Connection::open(
+            path,
+            OpenFlags::SQLITE_OPEN_READ_WRITE
+                | OpenFlags::SQLITE_OPEN_CREATE
+                | OpenFlags::SQLITE_OPEN_URI
+                | OpenFlags::SQLITE_OPEN_NO_MUTEX,
+            &INJECTOR_METHODS,
+            // safety: hook is dropped after connection
+            ctx,
+            DEFAULT_AUTO_CHECKPOINT,
+        )?;
+
+        connection.execute("CREATE TABLE IF NOT EXISTS libsql_temp_injection (x)", ())?;
+
+        Ok(Self {
+            is_txn: false,
+            buffer,
+            capacity: buffer_capacity,
+            connection: Arc::new(Mutex::new(connection)),
+        })
+    }
+
+    /// Inject on frame into the log. If this was a commit frame, returns Ok(Some(FrameNo)).
+    pub(crate) fn inject_frame(&mut self, frame: Frame) -> crate::Result<Option<FrameNo>> {
+        let frame_close_txn = frame.header().size_after != 0;
+        self.buffer.lock().push_back(frame);
+        if frame_close_txn || self.buffer.lock().len() >= self.capacity {
+            if !self.is_txn {
+                self.begin_txn()?;
+            }
+            return self.flush();
+        }
+
+        Ok(None)
+    }
+
+    /// Flush the buffer to libsql WAL.
+    /// Trigger a dummy write, and flush the cache to trigger a call to xFrame. The buffer's frame
+    /// are then injected into the wal.
+    fn flush(&mut self) -> crate::Result<Option<FrameNo>> {
+        let lock = self.buffer.lock();
+        // the frames in the buffer are either monotonically increasing (log) or decreasing
+        // (snapshot). Either way, we want to find the biggest frameno we're about to commit, and
+        // that is either the front or the back of the buffer
+        let last_frame_no = match lock.back().zip(lock.front()) {
+            Some((b, f)) => f.header().frame_no.max(b.header().frame_no),
+            None => {
+                tracing::trace!("nothing to inject");
+                return Ok(None);
+            }
+        };
+
+        drop(lock);
+
+        let connection = self.connection.lock();
+        connection.execute("INSERT INTO libsql_temp_injection VALUES (42)", ())?;
+        // force call to xframe
+        match connection.cache_flush() {
```

Comment:
@MarinPostma did you observe `xFrames` **not** being called without it? I found it weird, the docs https://www.sqlite.org/c3ref/db_cacheflush.html suggest that it's for flushing pages mid-transaction, and a single insert is a whole transaction on its own. It should be flushed to disk the moment you finish calling `execute`. And rusqlite's `cache_flush` is just a wrapper over `sqlite3_db_cacheflush`.

Reply:
> we are in the context of a transaction here

Ok, that's new, and proves I didn't follow the idea closely enough.. I see the `begin_txn` now. And I suppose we take it so that we can inject frames in multiple xFrames calls, but we need it to happen atomically anyway?

Reply:
Yea sir, this way we can apply very large transactions, such as snapshots